File: 0005_Longest_Palindromic_Substring/Solution2.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        res = ''
        for i in range(len(s)):
            # case 1 abcba
            left, right = i, i
            while 0 <= left and right < len(s) and s[left] == s[right]:
                new_str = s[left:right+1]
                if len(new_str) > len(res):
                    res = new_str
                left -= 1
                right += 1
            # case 2 abba
            left, right = i, i + 1
            while 0 <= left and right < len(s) and s[left] == s[right]:
                new_str = s[left:right+1]
                if len(new_str) > len(res):
                    res = new_str
                left -= 1
                right += 1
        return res

def test(test_name, s, expected):
    res = Solution().longestPalindrome(s)
    if res in expected:
        print(test_name + ' success.')
    else:
        print(test_name + ' failed.')

if __name__ == "__main__":
    s1 = "babad"
    expected1 = ['bab', 'aba']
    test('tset1', s1, expected1)

    s2 = "cbbd"
    expected2 = ["bb"]
    test('test2', s2, expected2)

    s3 = ''
    expected3 = ['']
    test('test3', s3, expected3)

File: 0367_Valid_Perfect_Square/Solution2.cpp
#include <cstdio>
#include <iostream>
using namespace std;

class Solution {
public:
    bool isPerfectSquare(int num) {
        // r set plus 1, for num = 1
        int l = 1, r = num / 2 + 1;
        int mid;
        long sq;
        while (l <= r) {
            mid = l + (r - l) / 2;
            sq = (long)mid * mid;   // convert long then compute, otherwise overflow happens
            if (sq < num)
                l = mid + 1;
            else if (sq > num)
                r = mid - 1;
            else
                return true;
        }
        return false;
    }
};

void test(string test_name, int num, bool expected)
{
    bool res = Solution().isPerfectSquare(num);
    if (res == expected)
        printf("%s succeed\n", test_name.c_str());
    else
        printf("%s fail\n", test_name.c_str());
}

int main()
{
    int num1 = 16;
    bool expected1 = true;
    test("test1", num1, expected1);

    int num2 = 14;
    bool expected2 = false;
    test("test2", num2, expected2);

    int num3 = 1;
    bool expected3 = true;
    test("test3", num3, expected3);

    int num4 = 2000105819;
    bool expected4 = false;
    test("test4", num4, expected4);

    return 0;
}

